text/0005-coretime-interface.md

```diff
@@ -0,0 +1,183 @@
+- The interface MUST allow for Instantaneous Coretime Market Credits to be communicated.
+- The interface MUST allow for the allocating chain to instruct changes to the number of cores which it is able to allocate.
+- The interface MUST allow for the allocating chain to be notified of changes to the number of cores which are able to be allocated by the allocating chain.
+
+## Stakeholders
+
+Primary stakeholder sets are:
+
+- Developers of the Relay-chain core-management logic.
+- Developers of the Brokerage System Chain and its pallets.
+
+_Socialization:_
+
+This content of this RFC was discussed in the Polkdot Fellows channel.
+
+## Explanation
+
+The interface has two sections: The messages which the Relay-chain is able to receive from the allocating parachain (the *UMP message types*), and messages which the Relay-chain is able to send to the allocating parachain (the *DMP message types*). These messages are expected to be able to be implemented in a well-known pallet and called with the XCM `Transact` instruction.
+
+Future work may include these messages being introduced into the XCM standard.
+
+### UMP Message Types
+
+#### `request_core_count`
+
+Prototype:
+
+```
+fn request_core_count(
+    count: u16,
+)
+```
+
+Requests the Relay-chain to alter the number of schedulable cores to `count`. Under normal operation, the Relay-chain SHOULD send a `notify_core_count(count)` message back.
+
+#### `request_revenue_info_at`
+
+Prototype:
+
+```
+fn request_revenue_at(
+    when: BlockNumber,
+)
+```
+
+Requests that the Relay-chain send a `notify_revenue` message back at or soon after Relay-chain block number `when` whose `until` parameter is equal to `when`.
+
+The period in to the past which `when` is allowed to be may be limited; if so the limit should be understood on a channel outside of this proposal. In the case that the request cannot be serviced because `when` is too old a block then a `notify_revenue` message must still be returned, but its `revenue` field may be `None`.
+
+#### `credit_account`
+
+Prototype:
+
+```
+fn credit_account(
+    who: AccountId,
+    amount: Balance,
+)
+```
+
+Instructs the Relay-chain to add the `amount` of DOT to the Instantaneous Coretime Market Credit account of `who`.
+
+It is expected that Instantaneous Coretime Market Credit on the Relay-chain is NOT transferrable and only redeemable when used to assign cores in the Instantaneous Coretime Pool.
+
+#### `assign_core`
+
+Prototype:
+
+```
+type PartsOf57600 = u16;
+enum CoreAssignment {
+    InstantaneousPool,
+    Task(ParaId),
+}
+fn assign_core(
+    core: CoreIndex,
+    begin: BlockNumber,
+    assignment: Vec<(CoreAssignment, PartsOf57600)>,
+    end_hint: Option<BlockNumber>,
+)
+```
+
+Requirements:
+
+```
+assert!(core < core_count);
+assert!(targets.iter().map(|x| x.0).is_sorted());
```

Comment:
They don't strictly need to be, but it's better to start with a more constrained interface

Reply:
I mean it will depend on the implementation of how to sort `CoreAssignment`, but this will probably mean that lower task numbers will be scheduled earlier?

Reply:
I really don't get the need for sorting it. Especially when we want to express something like `Task(1), Task(2)` should both use 1/2 of the same core at the same block.

Reply:
i have no opinion at all here.

Reply:
Given the comments from @rphmeier below, I see why the sorting doesn't has any effect